go/store/nbs/generational_chunk_store.go

```diff
@@ -0,0 +1,302 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nbs
+
+import (
+	"context"
+	"io"
+
+	"github.com/dolthub/dolt/go/store/chunks"
+	"github.com/dolthub/dolt/go/store/hash"
+)
+
+var _ chunks.ChunkStore = (*GenerationalNBS)(nil)
+var _ chunks.ChunkStoreGarbageCollector = (*GenerationalNBS)(nil)
+var _ chunks.GenerationalCS = (*GenerationalNBS)(nil)
+var _ TableFileStore = (*GenerationalNBS)(nil)
+
+type GenerationalNBS struct {
+	oldGen *NomsBlockStore
+	newGen *NomsBlockStore
+}
+
+func NewGenerationalCS(oldGen, newGen *NomsBlockStore) *GenerationalNBS {
+	if oldGen.Version() != newGen.Version() {
+		panic("oldgen and newgen chunkstore versions vary")
+	}
+
+	return &GenerationalNBS{
+		oldGen: oldGen,
+		newGen: newGen,
+	}
+}
+
+func (gcs *GenerationalNBS) NewGen() chunks.ChunkStoreGarbageCollector {
+	return gcs.newGen
+}
+
+func (gcs *GenerationalNBS) OldGen() chunks.ChunkStoreGarbageCollector {
+	return gcs.oldGen
+}
+
+// Get the Chunk for the value of the hash in the store. If the hash is absent from the store EmptyChunk is returned.
+func (gcs *GenerationalNBS) Get(ctx context.Context, h hash.Hash) (chunks.Chunk, error) {
+	c, err := gcs.oldGen.Get(ctx, h)
```

Comment:
I wonder about looking in `oldGen` before `newGen`. Presumably, `newGen` is much smaller than `oldGen`. 

A lookup miss in `newGen` followed by a successful lookup in `oldGen` would be faster than a lookup miss in `oldGen` followed by a successful lookup in `newGen`. What's not clear is the relative lookup frequency in new vs old. If the `newGen` contains any chunks, one of them must be the dataset map, and in that case chunks in `oldGen` must be accessed through refs stored in chunks in `newGen`.

My guess is that it's probably faster overall to check `newGen` first